### PR TITLE
Allow Library Templates to be associated with multiple projects

### PR DESCRIPTION
**Pull request by Kashif-S**
Jira: GLT-2439

### CODE REVIEW

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java [hunk 1]
```diff
@@ -2313,7 +2313,7 @@ public static LibraryTemplateDto asDto(LibraryTemplate from) {
 
     dto.setId(from.getId());
     dto.setAlias(from.getAlias());
-    dto.setProjectId(from.getProject().getId());
+    dto.setProjectIds(from.getProjects().stream().map(project -> project.getId()).collect(Collectors.toList()));
```
Comment (sqwci) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java]: MINOR Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java [hunk 2]
```diff
@@ -687,10 +688,14 @@ protected LibraryDto createDtoFromParent(Sample item) {
...
-          .map(projectId -> templateService.listLibraryTemplatesForProject(projectId))
-          .filter(list -> !list.isEmpty())
-          .map(Dtos::asLibraryTemplateDtos)
-          .collect(Collectors.toMap(list -> list.get(0).getProjectId(), list -> list));
+          .map(projectId -> {
+            Map<Long, List<LibraryTemplateDto>> map = new HashMap<>();
+            map.put(projectId, Dtos.asLibraryTemplateDtos(templateService.listLibraryTemplatesForProject(projectId)));
+            return map;
+          })
+          .filter(map -> !map.values().stream().allMatch(value -> value.isEmpty()))
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java]: MINOR Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

**Comment by sqwci** [comment]
SonarQube analysis reported 30 issues
* ![CRITICAL][CRITICAL] 9 critical
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 20 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L543](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L543): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L569](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L569): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L621](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L621): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L830](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L830): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L848](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L848): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L871](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L871): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L933](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L933): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L948](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L948): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L993](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L993): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![MAJOR][MAJOR] [EditLibraryController.java#L456](https://github.com/TGAC/miso-lims/blob/9d3c9437afd555d64f57e9f7c862b8f50280da35/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L456): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java (hunk 1 above)
Comment (apmasell) [reply miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java]: Yes, please.

File: sqlstore/src/main/resources/db/migration/V8812__LibraryTemplate_Multiple_Projects.sql [hunk 3]
```diff
@@ -0,0 +1,18 @@
+DROP TABLE IF EXISTS `LibraryTemplate_Project`;
+
+CREATE TABLE `LibraryTemplate_Project` (
+  `LibraryTemplate_libraryTemplateId` bigint(20) NOT NULL,
+  `Project_projectId` bigint(20) NOT NULL,
```
Comment (djcooke) [inline sqlstore/src/main/resources/db/migration/V8812__LibraryTemplate_Multiple_Projects.sql]: Please rename to `libraryTemplateId` and `projectId`. We've left the older fields named this way, but have been getting away from those long foreign key field names

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/LibraryTemplate.java [hunk 4]
```diff
@@ -97,12 +101,12 @@ public void setAlias(String alias) {
...
 
-  public Project getProject() {
-    return project;
+  public List<Project> getProjects() {
+    return projects;
   }
 
-  public void setProject(Project project) {
-    this.project = project;
+  public void setProject(List<Project> projects) {
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/LibraryTemplate.java]: `setProjects`*

**Comment by Kashif-S** [comment]
I was planning on updating the PR for the front-end Library Template creation to add the add projects feature after this one gets merged.

**Comment by sqwci** [comment]
SonarQube analysis reported 9 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L549](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L549): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L575](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L575): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L628](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L628): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L854](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L854): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L902](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L902): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L970](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L970): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [EditLibraryController.java#L462](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L462): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MINOR][MINOR] [EditLibraryController.java#L401](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L401): Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [EditLibraryController.java#L706](https://github.com/TGAC/miso-lims/blob/65823cb98b327e373430f2ed86bd025b6ef6c2f8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L706): Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**